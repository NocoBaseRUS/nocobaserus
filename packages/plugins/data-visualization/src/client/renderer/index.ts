export * from './ChartLibrary';
export * from './ChartRenderer';
export * from './ChartRendererProvider';
export * from './library';
