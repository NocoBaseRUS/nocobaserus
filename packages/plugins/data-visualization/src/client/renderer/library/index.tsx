import { AntdLibrary } from './AntdLibrary';
import { G2PlotLibrary } from './G2PlotLibrary';

export const InternalLibrary = { ...G2PlotLibrary, ...AntdLibrary };
