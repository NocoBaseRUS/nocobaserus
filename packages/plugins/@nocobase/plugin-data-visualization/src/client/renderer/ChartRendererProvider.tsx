import { css } from '@emotion/css';
import { useFieldSchema } from '@formily/react';
import { MaybeCollectionProvider, useAPIClient, useRequest } from '@nocobase/client';
import React, { createContext, useContext, useEffect } from 'react';
import { parseField } from '../utils';
import { ChartDataContext } from '../block/ChartDataProvider';

export type MeasureProps = {
  field: string | string[];
  aggregation?: string;
  alias?: string;
};

export type DimensionProps = {
  field: string | string[];
  alias?: string;
  format?: string;
};

export type TransformProps = {
  field: string;
  type: string;
  format: string;
};

export type QueryProps = Partial<{
  measures: MeasureProps[];
  dimensions: DimensionProps[];
  orders: {
    field: string;
    order: 'asc' | 'desc';
  }[];
  filter: any;
  limit: number;
  sql: {
    fields?: string;
    clauses?: string;
  };
}>;

export type ChartRendererProps = {
  collection: string;
  query?: QueryProps;
  config?: {
    chartType: string;
    general: any;
    advanced: any;
  };
  transform?: TransformProps[];
  mode?: 'builder' | 'sql';
  configuring?: boolean;
};

export const ChartRendererContext = createContext<
  {
    service: any;
    data?: any[];
  } & ChartRendererProps
>({} as any);

export const ChartRendererProvider: React.FC<ChartRendererProps> = (props) => {
  const { query, config, collection, transform, configuring } = props;
  const { addChart } = useContext(ChartDataContext);
  const schema = useFieldSchema();
  const api = useAPIClient();
  const service = useRequest(
    (collection, query) =>
      new Promise((resolve, reject) => {
        if (!(collection && query?.measures?.length)) return resolve(undefined);
        api
          .request({
            url: 'charts:query',
            method: 'POST',
            data: {
              uid: schema?.['x-uid'],
              collection,
              ...query,
              dimensions: (query?.dimensions || []).map((item: DimensionProps) => {
                const dimension = { ...item };
                if (item.format && !item.alias) {
                  const { alias } = parseField(item.field);
                  dimension.alias = alias;
                }
                return dimension;
              }),
              measures: (query?.measures || []).map((item: MeasureProps) => {
                const measure = { ...item };
                if (item.aggregation && !item.alias) {
                  const { alias } = parseField(item.field);
                  measure.alias = alias;
                }
                return measure;
              }),
            },
          })
          .then((res) => {
            resolve(res?.data?.data);
          })
          .catch(reject);
      }),
    {
      defaultParams: [collection, query],
    },
  );

  useEffect(() => {
    if (!configuring && schema?.['x-uid']) {
      addChart(schema?.['x-uid'], { collection, service, query });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [collection]);

  return (
    <MaybeCollectionProvider collection={collection}>
      <div
        className={css`
          .ant-card {
            box-shadow: none;
          }
        `}
      >
        <ChartRendererContext.Provider value={{ collection, config, transform, service, query }}>
          {props.children}
        </ChartRendererContext.Provider>
      </div>
    </MaybeCollectionProvider>
  );
};
