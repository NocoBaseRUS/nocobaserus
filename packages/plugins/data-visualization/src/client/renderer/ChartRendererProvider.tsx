import { css } from '@emotion/css';
import { MaybeCollectionProvider } from '@nocobase/client';
import React, { createContext } from 'react';

export type MeasureProps = {
  field: string | string[];
  aggregation?: string;
  alias?: string;
};

export type DimensionProps = {
  field: string | string[];
  alias?: string;
  format?: string;
};

export type TransformProps = {
  field: string;
  type: string;
  format: string;
};

export type QueryProps = Partial<{
  measures: MeasureProps[];
  dimensions: DimensionProps[];
  orders: {
    field: string;
    order: 'asc' | 'desc';
  }[];
  filter: any;
  limit: number;
  sql: {
    fields?: string;
    clauses?: string;
  };
}>;

export type ChartRendererProps = {
  collection: string;
  query?: QueryProps;
  config?: {
    chartType: string;
    general: any;
    advanced: any;
  };
  transform?: TransformProps[];
  mode?: 'builder' | 'sql';
};

export const ChartRendererContext = createContext<ChartRendererProps>({} as any);

export const ChartRendererProvider: React.FC<ChartRendererProps> = (props) => {
  const { collection } = props;
  return (
    <MaybeCollectionProvider collection={collection}>
      <div
        className={css`
          .ant-card {
            box-shadow: none;
          }
          .ant-card-body {
            padding: 0;
          }
        `}
      >
        <ChartRendererContext.Provider value={{ ...props }}>{props.children}</ChartRendererContext.Provider>
      </div>
    </MaybeCollectionProvider>
  );
};
