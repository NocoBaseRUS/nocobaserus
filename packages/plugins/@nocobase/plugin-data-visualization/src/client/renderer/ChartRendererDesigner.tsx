/**
 * This file is part of the NocoBase (R) project.
 * Copyright (c) 2020-2024 NocoBase Co., Ltd.
 * Authors: NocoBase Team.
 *
 * This project is dual-licensed under AGPL-3.0 and NocoBase Commercial License.
 * For more information, please refer to: https://www.nocobase.com/agreement.
 */

import { useField, useFieldSchema } from '@formily/react';
import {
  GeneralSchemaDesigner,
  SchemaSettingsDivider,
  SchemaSettingsItem,
  SchemaSettingsRemove,
  SchemaSettingsSelectItem,
  gridRowColWrap,
  useCollection_deprecated,
  useDataSource,
  useDesignable,
} from '@nocobase/client';
import React, { useContext } from 'react';
import { ChartConfigContext } from '../configure';
import { useChartsTranslation } from '../locale';
import { createRendererSchema } from '../utils';
import { ChartRendererContext } from './ChartRendererProvider';
import { ChartDataContext } from '../block/ChartDataProvider';

export const SchemaSettingsAutoRefresh: React.FC = () => {
  const { t } = useChartsTranslation();
  const field = useField();
  const fieldSchema = useFieldSchema();
  const { dn } = useDesignable();
  return (
    <SchemaSettingsSelectItem
      title={t('Auto refresh')}
      value={field.decoratorProps?.config?.autoRefresh || false}
      onChange={(v) => {
        const config = {
          ...field.decoratorProps?.config,
          autoRefresh: v,
        };
        field.decoratorProps.config = config;
        fieldSchema['x-decorator-props'] = field.decoratorProps;
        dn.emit('patch', {
          schema: {
            ['x-uid']: fieldSchema['x-uid'],
            'x-decorator-props': field.decoratorProps,
          },
        });
        dn.refresh();
      }}
      options={[
        {
          label: t('Off'),
          value: false,
        },
        {
          label: '5s',
          value: 5,
        },
        {
          label: '10s',
          value: 10,
        },
        {
          label: '30s',
          value: 30,
        },
        {
          label: '1m',
          value: 60,
        },
        {
          label: '5m',
          value: 300,
        },
        {
          label: '15m',
          value: 900,
        },
        {
          label: '30m',
          value: 1800,
        },
        {
          label: '1h',
          value: 3600,
        },
        {
          label: '2h',
          value: 7200,
        },
        {
          label: '1d',
          value: 86400,
        },
      ]}
    />
  );
};

export function ChartRendererDesigner() {
  const { t } = useChartsTranslation();
  const { setVisible, setCurrent } = useContext(ChartConfigContext);
  const { removeChart } = useContext(ChartDataContext);
  const { service } = useContext(ChartRendererContext);
  const field = useField();
  const schema = useFieldSchema();
  const { insertAdjacent } = useDesignable();
  const dataSource = useDataSource();
  const { name, title } = useCollection_deprecated();
  return (
    <GeneralSchemaDesigner disableInitializer title={title || name}>
      <SchemaSettingsItem
        title="Configure"
        key="configure"
        onClick={async () => {
          setCurrent({ schema, field, dataSource: dataSource.key, collection: name, service, data: service.data });
          setVisible(true);
        }}
      >
        {t('Configure')}
      </SchemaSettingsItem>
      <SchemaSettingsItem
        title="Duplicate"
        key="duplicate"
        onClick={() => insertAdjacent('afterEnd', gridRowColWrap(createRendererSchema(schema?.['x-decorator-props'])))}
      >
        {t('Duplicate')}
      </SchemaSettingsItem>
      <SchemaSettingsAutoRefresh />
      {/* <SchemaSettingsBlockTitleItem /> */}
      <SchemaSettingsDivider />
      <SchemaSettingsRemove
        // removeParentsIfNoChildren
        breakRemoveOn={{
          'x-component': 'ChartV2Block',
        }}
        confirm={{
          onOk: () => {
            removeChart(schema['x-uid']);
          },
        }}
      />
    </GeneralSchemaDesigner>
  );
}
