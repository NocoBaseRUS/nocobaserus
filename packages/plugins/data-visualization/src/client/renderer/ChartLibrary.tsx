import { ISchema } from '@formily/react';
import React, { createContext, useContext } from 'react';
import { FieldOption } from '../hooks';
import { lang } from '../locale';
import { parseField } from '../utils';
import { QueryProps } from './ChartRendererProvider';

/**
 * @params {usePropsFunc} useProps - Accept the information that the chart component needs to render,
 * process it and return the props of the chart component.
 */
export type usePropsFunc = (props: {
  data: any[];
  fieldProps: {
    [field: string]: FieldOption & {
      transformer: (val: any) => string;
    };
  };
  general: any;
  advanced: any;
}) => any;

export type ChartProps = {
  name: string;
  component: React.FC<any>;
  schema?: ISchema;
  useProps?: usePropsFunc;
  // The init function is used to initialize the configuration of the chart component from the query configuration.
  init?: (
    fields: FieldOption[],
    query: {
      measures?: QueryProps['measures'];
      dimensions?: QueryProps['dimensions'];
    },
  ) => {
    general?: any;
    advanced?: any;
  };
  reference?: {
    title: string;
    link: string;
  };
};

export type Charts = {
  [type: string]: ChartProps;
};

export type ChartLibraries = {
  [library: string]: {
    enabled: boolean;
    charts: Charts;
  };
};

export const ChartLibraryContext = createContext<ChartLibraries>({});

export const useCharts = (): Charts => {
  const library = useContext(ChartLibraryContext);
  return Object.values(library)
    .filter((l) => l.enabled)
    .reduce((charts, l) => ({ ...charts, ...l.charts }), {});
};

export const useChartTypes = (): {
  label: string;
  children: (ChartProps & {
    key: string;
    label: string;
    value: string;
  })[];
}[] => {
  const library = useContext(ChartLibraryContext);
  return Object.entries(library)
    .filter(([_, l]) => l.enabled)
    .reduce((charts, [name, l]) => {
      const children = Object.entries(l.charts).map(([type, chart]) => ({
        ...chart,
        key: type,
        label: chart.name,
        value: type,
      }));
      return [
        ...charts,
        {
          label: lang(name),
          children,
        },
      ];
    }, []);
};

export const useToggleChartLibrary = () => {
  const ctx = useContext(ChartLibraryContext);
  return {
    toggle: (library: string) => {
      ctx[library].enabled = !ctx[library].enabled;
    },
  };
};

export const ChartLibraryProvider: React.FC<{
  name: string;
  charts: Charts;
}> = (props) => {
  const { children, charts, name } = props;
  const ctx = useContext(ChartLibraryContext);
  const library = {
    ...ctx,
    [name]: {
      charts,
      enabled: true,
    },
  };
  return <ChartLibraryContext.Provider value={library}>{children}</ChartLibraryContext.Provider>;
};

export const infer = (
  fields: FieldOption[],
  {
    measures,
    dimensions,
  }: {
    measures?: QueryProps['measures'];
    dimensions?: QueryProps['dimensions'];
  },
) => {
  let xField: FieldOption;
  let yField: FieldOption;
  let seriesField: FieldOption;
  let yFields: FieldOption[];
  const getField = (fields: FieldOption[], selected: { field: string | string[]; alias?: string }) => {
    if (selected.alias) {
      return fields.find((f) => f.value === selected.alias);
    }
    const { alias } = parseField(selected.field);
    return fields.find((f) => f.value === alias);
  };
  if (measures?.length) {
    yField = getField(fields, measures[0]);
    yFields = measures.map((m) => getField(fields, m));
  }
  if (dimensions) {
    if (dimensions.length === 1) {
      xField = getField(fields, dimensions[0]);
    } else if (dimensions.length > 1) {
      // If there is a time field, it is used as the x-axis field by default.
      let xIndex: number;
      dimensions.forEach((d, i) => {
        const field = getField(fields, d);
        if (['date', 'time', 'datetime'].includes(field?.type)) {
          xField = field;
          xIndex = i;
        }
      });
      if (xIndex) {
        // If there is a time field, the other field is used as the series field by default.
        const index = xIndex === 0 ? 1 : 0;
        seriesField = getField(fields, dimensions[index]);
      } else {
        xField = getField(fields, dimensions[0]);
        seriesField = getField(fields, dimensions[1]);
      }
    }
  }
  return { xField, yField, seriesField, yFields };
};

export const commonInit: ChartProps['init'] = (fields, { measures, dimensions }) => {
  const { xField, yField, seriesField } = infer(fields, { measures, dimensions });
  return {
    general: {
      xField: xField?.value,
      yField: yField?.value,
      seriesField: seriesField?.value,
    },
  };
};
