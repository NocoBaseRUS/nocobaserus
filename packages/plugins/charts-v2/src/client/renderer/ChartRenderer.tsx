import { useField, useFieldSchema } from '@formily/react';
import {
  GeneralSchemaDesigner,
  gridRowColWrap,
  SchemaSettings,
  useAPIClient,
  useCollection,
  useDesignable,
  useRequest,
} from '@nocobase/client';
import { Empty, Result, Typography } from 'antd';
import React, { useContext, useEffect, useState } from 'react';
import { ErrorBoundary } from 'react-error-boundary';
import { ChartConfigContext } from '../block';
import { useFieldsWithAssociation, useFieldTransformer } from '../hooks';
import { useChartsTranslation } from '../locale';
import { createRendererSchema, getField, parseField, processData } from '../utils';
import { useCharts } from './ChartLibrary';
import { ChartRendererContext, DimensionProps, MeasureProps, QueryProps } from './ChartRendererProvider';
const { Paragraph, Text } = Typography;

export const ChartRenderer: React.FC<{
  configuring?: boolean;
  runQuery?: any;
}> & {
  Designer: React.FC;
} = (props) => {
  const { t } = useChartsTranslation();
  const { setData: setQueryData, current } = useContext(ChartConfigContext);
  const { query, config, collection, transform } = useContext(ChartRendererContext);
  const { configuring, runQuery } = props;
  const general = config?.general || {};
  const advanced = config?.advanced || {};
  const schema = useFieldSchema();
  const currentSchema = schema || current?.schema;
  const fields = useFieldsWithAssociation(collection);
  const api = useAPIClient();
  const [data, setData] = useState<any[]>([]);
  const { runAsync } = useRequest(
    (query) =>
      api
        .request({
          url: 'charts:query',
          method: 'POST',
          data: {
            uid: currentSchema?.['x-uid'],
            collection,
            ...query,
            dimensions: (query?.dimensions || []).map((item: DimensionProps) => {
              const dimension = { ...item };
              if (item.format && !item.alias) {
                const { alias } = parseField(item.field);
                dimension.alias = alias;
              }
              return dimension;
            }),
            measures: (query?.measures || []).map((item: MeasureProps) => {
              const measure = { ...item };
              if (item.aggregation && !item.alias) {
                const { alias } = parseField(item.field);
                measure.alias = alias;
              }
              return measure;
            }),
          },
        })
        .then((res) => {
          const data = res?.data?.data || [];
          return processData(fields, data, { t });
        }),
    {
      manual: true,
      onSuccess: (data) => {
        setData(data);
      },
      onFinally(params, data, error: any) {
        if (!configuring) {
          return;
        }
        if (error) {
          const message = error?.response?.data?.errors?.map?.((error: any) => error.message).join('\n');
          setQueryData(message || error.message);
          return;
        }
        setQueryData(data);
      },
    },
  );

  useEffect(() => {
    setData([]);
    const run = async (query: QueryProps) => {
      if (
        query?.measures?.length
        // || (query?.sql?.fields && query?.sql?.clauses)
      ) {
        await runAsync(query);
      }
    };
    if (runQuery) {
      runQuery.current = run;
    }
    run(query);
  }, [query, runAsync, runQuery]);

  const charts = useCharts();
  const chart = charts[config?.chartType];
  const Component = chart?.component;
  const locale = api.auth.getLocale();
  const transformers = useFieldTransformer(transform, locale);
  const info = {
    data,
    general,
    advanced,
    fieldProps: Object.keys(data[0] || {}).reduce((props, name) => {
      if (!props[name]) {
        const field = getField(fields, name.split('.'));
        const transformer = transformers[name];
        props[name] = { ...field, transformer };
      }
      return props;
    }, {}),
    locale,
  };
  const componentProps = chart?.useProps?.(info) || info;
  const C = () =>
    Component ? (
      <ErrorBoundary
        onError={(error) => {
          console.error(error);
        }}
        FallbackComponent={ErrorFallback}
      >
        <Component {...componentProps} />
      </ErrorBoundary>
    ) : (
      <Empty image={Empty.PRESENTED_IMAGE_SIMPLE} description={t('Please configure chart')} />
    );

  return data && data.length ? (
    <C />
  ) : (
    <Empty image={Empty.PRESENTED_IMAGE_SIMPLE} description={t('Please configure and run query')} />
  );
};

ChartRenderer.Designer = function Designer() {
  const { t } = useChartsTranslation();
  const { setVisible, setCurrent } = useContext(ChartConfigContext);
  const field = useField();
  const schema = useFieldSchema();
  const { insertAdjacent } = useDesignable();
  const { name, title } = useCollection();
  return (
    <GeneralSchemaDesigner disableInitializer title={title || name}>
      <SchemaSettings.Item
        key="configure"
        onClick={() => {
          setCurrent({ schema, field, collection: name });
          setVisible(true);
        }}
      >
        {t('Configure')}
      </SchemaSettings.Item>
      <SchemaSettings.Item
        key="duplicate"
        onClick={() =>
          insertAdjacent('afterEnd', createRendererSchema(schema?.['x-decorator-props']), {
            wrap: gridRowColWrap,
          })
        }
      >
        {t('Duplicate')}
      </SchemaSettings.Item>
      <SchemaSettings.BlockTitleItem />
      <SchemaSettings.Divider />
      <SchemaSettings.Remove
        // removeParentsIfNoChildren
        breakRemoveOn={{
          'x-component': 'ChartV2Block',
        }}
      />
    </GeneralSchemaDesigner>
  );
};

const ErrorFallback = ({ error }) => {
  const { t } = useChartsTranslation();

  return (
    <div style={{ backgroundColor: 'white' }}>
      <Result status="error" title={t('Render Failed')} subTitle={t('Please check the configuration.')}>
        <Paragraph copyable>
          <Text type="danger" style={{ whiteSpace: 'pre-line', textAlign: 'center' }}>
            {error.message}
          </Text>
        </Paragraph>
      </Result>
    </div>
  );
};
