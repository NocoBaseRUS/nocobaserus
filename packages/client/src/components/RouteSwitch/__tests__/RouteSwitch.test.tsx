import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter as Router, useParams, Link } from 'react-router-dom';
import { RouteSwitch } from '../';

const components = {
  Index: () => <div>index</div>,
  Home: ({ children }) => (
    <div>
      <h1>Home</h1>
      {children}
    </div>
  ),
  Blog: ({ children }) => (
    <div>
      <h1>Blog</h1>
      {children}
    </div>
  ),
  BlogPost: () => {
    let { slug } = useParams<any>();
    return <div>Now showing post {slug}</div>;
  },
  Login: () => <div>login</div>,
  Register: () => <div>register</div>,
};

const routes = [
  {
    type: 'redirect',
    from: '/blog/123',
    to: '/blog/1234',
  },
  // {
  //   component: 'Home',
  //   routes: [
  //     {
  //       path: '/blog/123555',
  //       component: () => <div>/blog/123555</div>,
  //     },
  //   ],
  // },
  {
    path: '/blog',
    component: 'Blog',
    routes: [
      {
        path: '/blog/:slug',
        // exact: true,
        component: 'BlogPost',
      },
    ],
  },
  {
    component: 'Home',
    routes: [
      // {
      //   path: '/blog/123555',
      //   component: () => <div>/blog/123555</div>,
      // },
      {
        path: '/login',
        component: 'Login',
      },
      {
        path: '/register',
        component: 'Register',
      },
      {
        path: '/',
        // exact: true,
        component: 'Index',
      },
    ],
  },
];

it('route component', () => {
  const t = renderer
    .create(
      <Router initialEntries={['/']}>
        <RouteSwitch
          routes={[
            {
              path: '/',
              component: () => <div>test</div>,
            },
          ]}
        />
      </Router>,
    )
    .toJSON();
  expect(t).toMatchSnapshot();
});

it('pathname=/', () => {
  const t = renderer
    .create(
      <Router initialEntries={['/']}>
        <RouteSwitch routes={routes} components={components} />
      </Router>,
    )
    .toJSON();
  expect(t).toMatchSnapshot();
});

it('pathname=/login', () => {
  const t = renderer
    .create(
      <Router initialEntries={['/login']}>
        <RouteSwitch routes={routes} components={components} />
      </Router>,
    )
    .toJSON();
  expect(t).toMatchSnapshot();
});

it('pathname=/blog/123', () => {
  const t = renderer
    .create(
      <Router initialEntries={['/blog/123']}>
        <RouteSwitch routes={routes} components={components} />
      </Router>,
    )
    .toJSON();
  expect(t).toMatchSnapshot();
});
