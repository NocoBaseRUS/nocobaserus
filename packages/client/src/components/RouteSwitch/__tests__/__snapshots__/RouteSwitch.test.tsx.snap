// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pathname=/ 1`] = `
<div>
  <h1>
    Home
  </h1>
  <div>
    index
  </div>
</div>
`;

exports[`pathname=/blog/123 1`] = `
<div>
  <h1>
    Blog
  </h1>
  <div>
    Now showing post 
    1234
  </div>
</div>
`;

exports[`pathname=/login 1`] = `
<div>
  <h1>
    Home
  </h1>
  <div>
    login
  </div>
</div>
`;

exports[`route component 1`] = `
<div>
  test
</div>
`;
